### PR TITLE
Add a new process_src() public helper to look in the source directory.

### PR DESCRIPTION
**Pull request by dburgener**
Update the docs to recommend this as the top choice for setting up lalrpop.

Prior to this commit, the book incorrectly stated that `process_root()` looked in `./src`, while the documentation comments in the code correctly identified it as parsing the current directory.  Switching the book to recommend `process_src()` addresses this inconsistency as well.

### CODE REVIEW

**Comment by dburgener** [comment]
closes #908

**Comment by dburgener** [comment]
Adding a new public API element isn't breaking, but would require bumping the minor version in post-1.0.  If we merge this, we might want to include it in a minor version bump for easy cargo version specification for users relying on the new api.

**Comment by Pat-Lafon** [comment]
Do you think there is a goal of deprecating `process_root` in favor of `process_src`? If I understand this, `process_src` is the happy/typical path, and if you want any path tweaking, you can use `Configuration`?

**Comment by dburgener** [comment]
> Do you think there is a goal of deprecating process_root in favor of process_src? If I understand this, process_src is the happy/typical path, and if you want any path tweaking, you can use Configuration?

I went into this thinking that would be the goal, but now I think they're both independently useful.  Namely, `process_root()` is helpful in a case like lalrpop, where multiple crates are built in the same workspace (so there's no `src` at the top level).  I think having both, but recommending `src` in the common case makes sense to me, because 1. src will hard fail (not find the lalrpop file) if someone should have used root, but using root instead of src will have subtler issues, as seen in the report.  2. I suspect that single-crate workspaces are (much?) more common.

**Comment by youknowone** [comment]
> 2.  I suspect that single-crate workspaces are (much?) more common.

single-crate workspace will be less common. But probably multi-crate workspaces which compile the same lalrpop file in multiple crate will be even less common. Multi-crate workspace which contains a single crate compiling a single lalrpop file still will use process_src.

**Comment by Pat-Lafon** [comment]
One more question. I wonder how much benefit there is in having this separation/indirection between users and `Configuration`.

The main reason for this in my mind would be if we wanted to make enough changes to `Configuration` such that we were deprecating it or some of it's methods in favor of a new api. Then we could make the swap in the implementation of `process_src` with less real world breakage.

On the other hand, I don't think a `Configuration::new().set_in_dir("./src").process().unwrap()` is the worst default. It does make it a little easier to understand why lalrpop may not be finding your grammars(if they are in `grammars` instead of `src` as a convoluted example). It would be a little more consistent in leading users to more advanced features like [emit_rerun_directives](https://docs.rs/lalrpop/latest/lalrpop/struct.Configuration.html#method.emit_rerun_directives) or [generate_in_source_tree](https://docs.rs/lalrpop/latest/lalrpop/struct.Configuration.html#method.generate_in_source_tree). (I don't really use the rest of the options but I guess they are there for a reason)

File: lalrpop/src/api/mod.rs
Comment (Pat-Lafon) [inline lalrpop/src/api/mod.rs:261]: Nit: Calling `Path::new` is a little unnecessary as it will be done automatically for `&str`.

**Review by Pat-Lafon — APPROVED** [review]
None of my comments should be construed as blocking this pr. Please move forward as you/the others see fit.

**Comment by dburgener** [comment]
> The main reason for this in my mind would be if we wanted to make enough changes to Configuration such that we were deprecating it or some of it's methods in favor of a new api. Then we could make the swap in the implementation of process_src with less real world breakage.

I think of it as mainly a convenience function.  As a user, I like to grab a default and just get it working in early development.  If the default works for me, I can keep it, and if not, I can always upgrade to the full `Configuration` version once I have my bearings in the project a little more.

> On the other hand, I don't think a Configuration::new().set_in_dir("./src").process().unwrap() is the worst default.

Yeah, it's fine.  If we want to go that way, I'd be happy to submit a PR updating the book's rec to this.  As I said above, I do think that recommending `src` as the default location seems like a better strategy than recommending the root as the default.  Youknowone pushed back on the asserting that single-crate workspaces are more common, and I don't have data to back that up, but either way, preferring `src` as the default fails hard if they got it wrong, and preferring `.` as the default may fail subtly, or simply perform worse.

If we're recommending `src` as the default, it's a little weird to only have the convenience function `process_root()`, and not `process_src()` as well, and as I said above, deprecating `process_root()` seems to remove something that does have a lot of value.

I guess in terms of moving forward, we should probably agree on:

1. What is our recommendation to new users in the quick start guide?  As stated earlier, I think we should recommend `src` as a first choice, but state both options.
2. Do we want convenience functions like `process_root()` and `process_src()`?  If so, which ones?  (And if not `process_root()`, do we want to deprecate, or just deemphasize?)

I'm happy to hear push-back.  If we add to the public API, and then change our minds about it later, that has costs, so getting it right initially is definitely worth some time and back and forth.

**Comment by Pat-Lafon** [comment]
> If we're recommending `src` as the default, it's a little weird to only have the convenience function `process_root()`, and not `process_src()` as well, and as I said above, deprecating `process_root()` seems to remove something that does have a lot of value.

Sure, I've both sorta come around on this. If it makes users lives easier then I'm willing to support the extra convenience function since it is simple enough. I think my point can be mostly addressed by having more documentation pointing from these functions to `Configuration`.

> I guess in terms of moving forward, we should probably agree on:
> 
>     1. What is our recommendation to new users in the quick start guide?  As stated earlier, I think we should recommend `src` as a first choice, but state both options.

Recommend `src` since it probably is the right default. Let users know the other two options to reach for when they fall off of this happy path.

>     2. Do we want convenience functions like `process_root()` and `process_src()`?  If so, which ones?  (And if not `process_root()`, do we want to deprecate, or just deemphasize?)

I guess there isn't a whole lot to be gained from deprecating `process_root` if this is what most users have. But encouraging people to use `process_src` seems worthwhile to me.

> I'm happy to hear push-back. If we add to the public API, and then change our minds about it later, that has costs, so getting it right initially is definitely worth some time and back and forth.

👍

**Comment by dburgener** [comment]
@Pat-Lafon I know your approval is here already, but we did have some back and forth and discussion since then.  Does this seem ready to merge from your perspective?  Would you like to wait for any additional chime-ins from other maintainers?

If we feel good about this, it might be nice to get the public API addition in a release that increments the minor version anyways, so we could merge it before 0.21.

**Comment by Pat-Lafon** [comment]
Yup lets merge this. I'll make a small doc suggestion in a separate pr.

**Pull request merged**
